packages/contracts/contracts/test/ValidatorPool.t.sol

```diff
@@ -0,0 +1,402 @@
+    MockL2OutputOracle mockOracle;
+    KromaPortal portal;
+
+    uint256 internal finalizationPeriodSeconds;
+
+    event Bonded(
+        address indexed submitter,
+        uint256 indexed outputIndex,
+        uint128 amount,
+        uint128 expiresAt
+    );
+
+    event BondIncreased(address indexed challenger, uint256 indexed outputIndex, uint128 amount);
+
+    event Unbonded(uint256 indexed outputIndex, address indexed recipient, uint128 amount);
+
+    function setUp() public override {
+        super.setUp();
+
+        finalizationPeriodSeconds = oracle.FINALIZATION_PERIOD_SECONDS();
+
+        address oracleAddress = address(oracle);
+        MockL2OutputOracle mockOracleImpl = new MockL2OutputOracle(pool, address(challenger));
+        vm.prank(multisig);
+        Proxy(payable(oracleAddress)).upgradeTo(address(mockOracleImpl));
+        mockOracle = MockL2OutputOracle(oracleAddress);
+
+        portal = new KromaPortal({
+            _l2Oracle: oracle,
+            _guardian: guardian,
+            _paused: false,
+            _config: systemConfig,
+            _zkMerkleTrie: ZKMerkleTrie(address(0))
+        });
+    }
+
+    function test_constructor_succeeds() external {
+        assertEq(address(pool.L2_ORACLE()), address(oracle));
+        assertEq(pool.TRUSTED_VALIDATOR(), trusted);
+        assertEq(pool.MIN_BOND_AMOUNT(), minBond);
+    }
+
+    function test_deposit_succeeds() public {
+        vm.prank(trusted);
+        pool.deposit{ value: minBond }();
+        assertEq(pool.balanceOf(trusted), minBond);
+        assertTrue(pool.isValidator(trusted));
+        assertEq(pool.validatorCount(), 1);
+
+        vm.prank(asserter);
+        pool.deposit{ value: minBond }();
+        assertEq(pool.balanceOf(asserter), minBond);
+        assertTrue(pool.isValidator(asserter));
+        assertEq(pool.validatorCount(), 2);
+    }
+
+    function test_deposit_alreadyValidator_succeeds() external {
+        test_deposit_succeeds();
+
+        uint256 count = pool.validatorCount();
+        address nextValidator = pool.nextValidator();
+        uint256 balance = pool.balanceOf(nextValidator);
+
+        vm.prank(nextValidator);
+        pool.deposit{ value: 1 }();
+        assertEq(pool.balanceOf(trusted), balance + 1);
+        assertTrue(pool.isValidator(trusted));
+        assertEq(pool.validatorCount(), count);
+    }
+
+    function test_deposit_insufficientBalances_reverts() external {
+        vm.deal(asserter, 0);
+        vm.prank(asserter);
+        vm.expectRevert();
+        pool.deposit{ value: minBond }();
+    }
+
+    function test_withdraw_loseValidatorEligibility_succeeds() external {
+        test_deposit_succeeds();
+
+        uint256 count = pool.validatorCount();
+        address nextValidator = pool.nextValidator();
+        uint256 balance = pool.balanceOf(nextValidator);
+
+        vm.prank(nextValidator);
+        pool.withdraw(1);
+        assertEq(pool.balanceOf(nextValidator), balance - 1);
```

Comment:
Can we check if the ETH balance under the `nextValidator` account is equal to 1? for checking actual withdraw to the target account.

Reply:
I don't think that assertion is necessary.
If the `next validator` has no deposits, the transaction will revert with `insufficient balance`.